### ISSUE TITLE
It is possible to run two different JMX exporters for the same kafka broker?

### ISSUE THREAD

**Issue by wrengifo** [prometheus/jmx_exporter#672]
Hi, I need to run a JMX exporter in parallel for a kafka broker that already has one running. I tried run as an additional agent unsuccesfully. Do you know if it is possible?

**Comment by dhoard** [comment]
You can only run one Java agent (`-javaagent`) per JVM. This is a JVM constraint.

**Comment by wrengifo** [comment]
> You can only run one Java agent (`-javaagent`) per JVM. This is a JVM constraint.

Thanks for the answer. I wonder if could be possible run it as independent http server for the same broker?

**Comment by dhoard** [comment]
You could do that provided remote JMX is allowed.

I am not sure what your trying to accomplish.

If there is already a jmx-exporter agent running why wouldn't you just scrape it?

**Comment by wrengifo** [comment]
> You could do that provided remote JMX is allowed.
> 
> I am not sure what your trying to accomplish.
> 
> If there is already a jmx-exporter agent running why wouldn't you just scrape it?

My goal is to have 2 Exporter each one in its specific port. We need to maintain some metrics in a certain port, meanwhile we are exposing another exporter in a different port using different config-file.yaml.

**Comment by dhoard** [comment]
Adding more detail / clarification to my answer.

Java allows more than one agent to run, but the `jmx_prometheus_javaagent` agent isn't designed to be run twice.

---

Java documentation (https://docs.oracle.com/javase/6/docs/api/java/lang/instrument/package-summary.html)...

```
On implementations with a command-line interface, an agent is started by adding this option to the command-line:

-javaagent:jarpath[=options]

jarpath is the path to the agent JAR file. options is the agent options. This switch may be used multiple times on the same command-line, thus creating multiple agents.
```
---

Test using simple `HelloWorld` application and two declarations of `jmx_prometheus_javaagent`...

```
java -javaagent:./jmx_prometheus_javaagent/target/jmx_prometheus_javaagent-0.16.2-SNAPSHOT.jar=8080:config1.yaml -javaagent:./jmx_prometheus_javaagent/target/jmx_prometheus_javaagent-0.16.2-SNAPSHOT.jar=8081:config2.yaml HelloWorld
```

Error...

```
Usage: -javaagent:/path/to/JavaAgent.jar=[host:]<port>:<yaml configuration file> Failed to register Collector of type BuildInfoCollector: jmx_exporter_build_info is already in use by another Collector of type BuildInfoCollector
```

---

I would not recommend running more than one agent against a Kafka broker since some metrics are implemented inline, resulting in more load on the broker.

**Comment by dhoard** [comment]
@wrengifo I believe so... though still think it's bad practice.

**Comment by RamakrishnaHande** [comment]
@wrengifo  Hi , good morning. How did you fix this issue ??

Thanks in advance

**Comment by RamakrishnaHande** [comment]
I have prometheus target at 8090 and I see them on graffana charts and I see http://localohost:8090/metrics gives all spark related metrics.

Any custom metrics , I am exposing via 11200, and in code that itself creates a Server at 11200, hence I want to scrape 11200 also 

so when I do 

http://localohost:11200/metrics, I should be able to query the metrics exposed at that port

**Comment by dhoard** [comment]
Seems like an odd use case to me that all metrics aren't available at http://localohost:8090/metrics.

In your code, if you are creating your own embedded exporter (non-default `CollectorRegistry`, `HTTPServer`, configuration, etc.) then you can do anything you want.

**Comment by RamakrishnaHande** [comment]
> Seems like an odd use case to me that all metrics aren't available at http://localohost:8090/metrics.
> 
> In your code, if you are creating your own embedded exporter (non-default `CollectorRegistry`, `HTTPServer`, configuration, etc.) then you can do anything you want.
> Could you please show me an example ?? I have an 

"  new HTTPServer.Builder()
      .withPort(metricExporterPort)
      .build()
      "

where metricExporterPort = 11200

I may want to change this one
> This is a different use case than running 2 exporters against the same JVM. I suggest creating a new issue should if you want to have more discussion so as not to hijack this different issue.

**Comment by dhoard** [comment]
Sure... Something like...

```
int port = 11200;
CollectorRegistry registry = new CollectorRegistry();;
HTTPServer httpServer = new HTTPServer.Builder()
            .withRegistry(registry)
            .withPort(port)
            // any other options you want to set
            .build();

```

**Comment by dhoard** [comment]
Closing as stale. Please reopen if you would like to continue the discussion.
